src/components/macro/macro-action-editor/macro-action-editor.component.ts

```diff
@@ -0,0 +1,113 @@
+import {Component, EventEmitter, Input, OnInit, Output, ViewChild } from '@angular/core';
+
+import {EditableMacroAction} from '../../../config-serializer/config-items/macro-action/EditableMacroAction';
+import {MacroAction, macroActionType} from '../../../config-serializer/config-items/macro-action/MacroAction';
+import {MacroKeyTabComponent} from './tab/macro-key';
+
+enum TabName {
+    Keypress,
+    Text,
+    Mouse,
+    Delay
+};
+
+@Component({
+    selector: 'macro-action-editor',
+    template: require('./macro-action-editor.component.html'),
+    styles: [ require('./macro-action-editor.component.scss') ],
+    host: { 'class': 'macro-action-editor' }
+})
+export class MacroActionEditorComponent implements OnInit {
+    @Input() macroAction: MacroAction;
+
+    @Output() save = new EventEmitter<MacroAction>();
+    @Output() cancel = new EventEmitter<void>();
+
+    @ViewChild('tab') selectedTab: any;
```

Comment:
any is fine here because each tab is quite unique, only some of them implement Tab and I didn't want to create a completely redundant class or interface for these as there are no shared properties or methods.